### PR TITLE
Add feature to offset start times for event types

### PR DESCRIPTION
**Pull request by rjackson**
## What does this PR do?

This PR adds a new "Offset start time" option to event types, which can be used to push time slots back from aligning perfectly with calendar hours / intervals.

For example, you can set up an event to serve the slots:

- 9:10am to 10:00am
- 10:10am to 11:00am
- 11:10am to 12:00pm

By configuring it with:

- Event length: 50 minutes
- Offset start time: 10 minutes

This feature was requested through #8336

**New field in the Event Types edit screen**

image

I wasn't quite sure where to put this field. It doesn't seem like a common option to show prominently on the "Event Setup" tab. I ended up feeling like the "Limits" tab is the best place for it, as its conceptually similar to before & after buffer times – just aligned to the calendar slot frequency instead of prior or future events.

**An example appointment page**

image

**Environment**: Staging(main branch)

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update
  - (if we want to add a page / note about it as a new feature?)

## How should this be tested?

- [ ] Unit tests (getSchedule.test.ts and slots.test.ts have been updated)
- [ ] Create a new event type and on its edit page
  - [ ] See the offset starts at 0
  - [ ] Change it and successfully save it to another value
  - [ ] Blank the field and save, see it reset back to 0
  - [ ] Try to set the field to a negative value, and be blocked by validation
  - [ ] Set it as required for your own testing
- [ ] View the appointment page, and see results appear as configured

## Checklist

- ~~I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)~~
- ~~My code doesn't follow the style guidelines of this project~~
- ~~I haven't performed a self-review of my own code and corrected any misspellings~~
- ~~I haven't commented my code, particularly in hard-to-understand areas~~
- ~~I haven't checked if my PR needs changes to the documentation~~
- ~~I haven't checked if my changes generate no new warnings~~
- ~~I haven't added tests that prove my fix is effective or that my feature works~~
- ~~I haven't checked if new and existing unit tests pass locally with my changes~~

Custom checklist additions:

- End-to-end tests: I couldn't run these locally, so I'm hoping the PR can verify these for me.

Bounty hunting: /claim #8336

### CODE REVIEW

**Comment by mxstbr** [comment]
This looks _amazing_, exactly what I need! Hoping this'll land soon and I can switch over to cal.com :rocket:

**Comment by regexyl** [comment]
This is a great add-on for the issue - though I feel like from a long-term UX perspective, we'd want to simplify it so the user sees either:

1. Before + After event buffer time, or
2. Offset start time

Slightly less customizable, but I'd feel users would probably want one or the other anyway, not both? I have a feeling it'll be a little confusing if someone saw all three fields at once.

**Comment by rjackson** [comment]
Great feedback @regexyl, thank you :)

I'm wondering if it might be simpler to reuse the buffer time fields, but add a toggle to change how we use them when calculating slots. Instead of skipping over a slot because its too close to the previous booking, we use the buffer times as the offset instead.

My only worry would be coming up with a label & description which clearly describes that difference in functionality to users.

Something like this?

image

**Comment by regexyl** [comment]
Ooh, that's a great idea! 

I haven't thought this through clearly, though I think there are a few more edge cases to account for. The current toggle factors in both before and after buffers into the meeting time, but what if the user just wants to offset the before buffer and leave the one after untouched? Should two toggles work then or would it be more confusing? 🤨

Defo feel like it'll be nice to gather more opinions to see how this should look, but I think for me the first iteration does the job well.

**Comment by CarinaWolli** [comment]
@Jaibles what's your opinion on that?

**Comment by Jaibles** [comment]
Is this going to be super confusing for bookings with durations set to times like 20 or 50 minutes?
CleanShot 2023-05-08 at 13 19 58@2x

**Comment by rjackson** [comment]
I like that second design. As an uncommon option, I agree that hiding it behind a toggle would be best – those who understand what it does can then opt-into it as required.

I'll look into getting the PR updated as per that second design, if we're all happy to do so?

**Comment by Jaibles** [comment]
> I'll look into getting the PR updated as per that second design, if we're all happy to do so?

Yes please, and thank you @rjackson  🙏

**Comment by rjackson** [comment]
All updated :) Here's how it's looking now:

State | Screenshot
--- | ---
Enabled | Screenshot 2023-05-08 at 17 52 56
Disabled | Screenshot 2023-05-08 at 17 53 00

The settings toggle is open by default if `offsetStart` has a value. If not, it's closed by default.

Unchecking the toggle also sets `offsetStart` to zero, effectively disabling it.

**Comment by mxstbr** [comment]
@Jaibles 

> "the idea of a meeting at 9.05am upsets me 😅"

🤣 Thanks for all the great input, can't wait to see this go live!

**Review by ciaranha — COMMENTED** [review]
Looks good to me! Thank you @rjackson 🙏. Will let someone from core who can oversee the code changes actually approve.

**Comment by mxstbr** [comment]
Is there any way I can help this land? :pray:

File: apps/web/components/eventtype/EventLimitsTab.tsx
Comment (joeauyeung) [inline apps/web/components/eventtype/EventLimitsTab.tsx:170]: I love this and I think this will help a lot.

**Review by joeauyeung — APPROVED** [review]
LGTM

Test cases
- Offset by 5 minutes
- Offset by 120 minutes
- With a combination of before & after buffer times

File: apps/web/test/lib/getSchedule.test.ts
Comment (joeauyeung) [inline apps/web/test/lib/getSchedule.test.ts:755]: @rjackson is this meant to be here? Causing a type error

Comment (rjackson) [reply apps/web/test/lib/getSchedule.test.ts:755]: Ah, looks like 1eeb91a793e577b10c3037a2158d7ad052659b20 removed the need to pass ctx. Hadn't spotted that when merging in upstream changes. Tests seem to be passing happily now :)

**Comment by mxstbr** [comment]
Let's :shipit:?! 😍

**Review by JeroenReumkens — APPROVED** [review]
Amazing work @rjackson. Really like the simplicity you were able to keep here 🙌
Final approve is up to you @zomars as a code owner.

**Comment by PeerRich** [comment]
tldr: waiting for @zomars for final review since this is touching booking logic we wanna be extra sure

**Comment by mxstbr** [comment]
Trevor Noah with an excited face saying "The suspense is killing meeeee"

**Comment by PeerRich** [comment]
preparing a merge. we can review after, too! looks good, been reviewed by two people

**Pull request merged**
